### PR TITLE
Seq2Seq model support for DPO

### PR DESCRIPTION
**Pull request by gaetanlop**
Draft PR following #579. The goal is to make DPO support encoder-decoder models like T5. 

Things to add:
- Improve the collator so that it also returns the text version of the prompt/chosen/rejected samples.
- Add decoder input ids to the collator for models that do not directly create them from labels.
- Find a better way to know if the model is an encoder-decoder than directly passing a param in init.
- Modify the dpo example if needed.

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -97,6 +97,8 @@ def __init__(
         preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
         max_length: Optional[int] = None,
         max_prompt_length: Optional[int] = None,
+        max_target_length: Optional[int] = None,
+        is_encoder_decoder: bool = None,
```
Comment (tomaarsen) [inline trl/trainer/dpo_trainer.py:101]:
```suggestion
```

Comment (tomaarsen) [reply trl/trainer/dpo_trainer.py:101]: You can use `model.config.is_encoder_decoder`:
https://github.com/huggingface/transformers/blob/main/src/transformers/configuration_utils.py#L277-L278
All Configs have this option.

Do note however that in this `__init__` we might have a `model_init` instead of a `model`, so we'll have to work around that.

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -128,13 +130,23 @@ def __init__(
                 )
                 max_prompt_length = 128
 
+            if max_target_length is None & is_encoder_decoder:
```
Comment (tomaarsen) [inline trl/trainer/dpo_trainer.py:133]:
```suggestion
            if max_target_length is None and is_encoder_decoder:
```
+ you can use `model.is_encoder_decoder` if `model` is passed.

File: trl/trainer/dpo_trainer.py [hunk 3]
```diff
@@ -189,24 +202,46 @@ def concatenated_inputs(self, batch: Dict[str, Union[List, torch.LongTensor]]) -
...
+                if k.startswith("rejected") and isinstance(batch[k], torch.Tensor):
+                    pad_value = self.label_pad_token_id if "labels" in k else self.padding_value
+                    concatenated_key = k.replace("rejected", "concatenated")
+                    concatenated_batch[concatenated_key] = torch.cat(
+                        (
+                            concatenated_batch[concatenated_key],
+                            pad_to_length(batch[k], max_length, pad_value=pad_value),
+                        ),
+                        dim=0,
+                    ).to(self.accelerator.device)
```
Comment (tomaarsen) [inline trl/trainer/dpo_trainer.py:244]:
```suggestion
        if self.is_encoder_decoder:
            max_length = max(batch["chosen_labels"].shape[1], batch["rejected_labels"].shape[1])
            pad_value = self.label_pad_token_id 
        else:
            max_length = max(batch["chosen_input_ids"].shape[1], batch["rejected_input_ids"].shape[1])
            pad_value = self.label_pad_token_id if "labels" in k else self.padding_value

        for k in batch:
            if k.startswith("chosen") and isinstance(batch[k], torch.Tensor):
                concatenated_key = k.replace("chosen", "concatenated")
                concatenated_batch[concatenated_key] = pad_to_length(batch[k], max_length, pad_value=pad_value)
        for k in batch:
            if k.startswith("rejected") and isinstance(batch[k], torch.Tensor):
                concatenated_key = k.replace("rejected", "concatenated")
                concatenated_batch[concatenated_key] = torch.cat(
                    (
                        concatenated_batch[concatenated_key],
                        pad_to_length(batch[k], max_length, pad_value=pad_value),
                    ),
                    dim=0,
                ).to(self.accelerator.device)
            if self.is_encoder_decoder:
                concatenated_batch["concatenated_input_ids"] = batch["prompt_input_ids"].repeat(2,1)
                concatenated_batch["concatenated_attention_mask"] = batch["prompt_attention_mask"].repeat(2,1)
```
I believe this is equivalent.

File: trl/trainer/dpo_trainer.py [hunk 4]
```diff
@@ -287,10 +323,19 @@ def concatenated_forward(
...
+            all_logits = model(
+                concatenated_batch["concatenated_input_ids"],
+                attention_mask=concatenated_batch["concatenated_attention_mask"],
+                labels = concatenated_batch["concatenated_labels"]
+            ).logits.to(torch.float32)
+        else:
+            all_logits = model(
+                concatenated_batch["concatenated_input_ids"],
+                attention_mask=concatenated_batch["concatenated_attention_mask"],
+            ).logits.to(torch.float32)
```
Comment (tomaarsen) [inline trl/trainer/dpo_trainer.py:337]:
```suggestion
        model_kwargs = {"labels": concatenated_batch["concatenated_labels"]} if self.is_encoder_decoder else {}
        all_logits = model(
            concatenated_batch["concatenated_input_ids"],
            attention_mask=concatenated_batch["concatenated_attention_mask"],
            **model_kwargs,
        ).logits.to(torch.float32)
```
We can simplify this too, but this is also a bit of a style preference.

File: trl/trainer/utils.py [hunk 5]
```diff
@@ -229,97 +231,124 @@ def tokenize_batch_element(
...
-                # adapted from https://stackoverflow.com/questions/73256206
-                if "prompt" in k:
-                    to_pad = [torch.LongTensor(ex[k][::-1]) for ex in batch]
-                else:
-                    to_pad = [torch.LongTensor(ex[k]) for ex in batch]
-                if k.endswith("_input_ids"):
+            if self.is_encoder_decoder:
+                to_pad = [torch.LongTensor(ex[k]) for ex in batch]
+            
+                if (k.startswith("prompt")) & (k.endswith("input_ids")):
```
Comment (tomaarsen) [inline trl/trainer/utils.py:320]:
```suggestion
                if k.startswith("prompt") and k.endswith("input_ids"):
```

File: trl/trainer/utils.py [hunk 6]
```diff
@@ -229,97 +231,124 @@ def tokenize_batch_element(
...
                     padding_value = self.tokenizer.pad_token_id
-                elif k.endswith("_labels"):
-                    padding_value = self.label_pad_token_id
                 elif k.endswith("_attention_mask"):
-                    padding_value = self.padding_value
+                    padding_value = 0
+                elif (k.startswith("chosen")) | (k.startswith("rejected")):
```
Comment (tomaarsen) [inline trl/trainer/utils.py:324]:
```suggestion
                elif k.startswith("chosen") or k.startswith("rejected"):
```

**Review by tomaarsen — COMMENTED** [review]
Hey! Thanks for starting this work. I wrote a few changes/simplifications that might be useful.

Another thing - this project uses `pre-commit`, which automatically runs some formatting scripts every time that you commit. You can use it with `pip install pre-commit` and `pre-commit install` to set it up (for this git repository only).
Otherwise the code quality CI will keep complaining.

If you're curious, this is what it runs: https://github.com/lvwerra/trl/blob/main/.pre-commit-config.yaml (p.s. maybe we can move to using ruff as well? It can replace isort).

**Comment by gaetanlop** [comment]
Thanks @tomaarsen I will incorporate that.

File: trl/trainer/dpo_trainer.py (hunk 3 above)
Comment (gaetanlop) [reply trl/trainer/dpo_trainer.py:244]: I don't think this will work as k is not defined at this point for the pad_value

File: trl/trainer/dpo_trainer.py (hunk 4 above)
Comment (gaetanlop) [reply trl/trainer/dpo_trainer.py:337]: Indeed, thanks!

File: trl/trainer/dpo_trainer.py (hunk 3 above)
Comment (tomaarsen) [reply trl/trainer/dpo_trainer.py]: Oh, whoops. I missed that variable.
That said, the pad_value can be moved down somewhat in my solution, and it'll still remove some duplicated code.

Comment (gaetanlop) [reply trl/trainer/dpo_trainer.py]: Yes I agree

File: trl/trainer/dpo_trainer.py (hunk 1 above)
Comment (gaetanlop) [reply trl/trainer/dpo_trainer.py]: Oh yes thanks I forget configs had this option

**Comment by gaetanlop** [comment]
The PR is still not ready, I need to add docstrings for the different arguments added and fix the code quality issue that comes from some added imports.

**Comment by tomaarsen** [comment]
You can try running `pre-commit run --all` and see if that makes the desired changes. For me it does. If that doesn't work, perhaps you can allow me to somehow push a commit to your branch and then I can resolve it for you, as I have the commit locally. I'm not quite sure how that works, though.

- Tom Aarsen

**Comment by gaetanlop** [comment]
Thanks a lot @tomaarsen it worked!

**Comment by kashif** [comment]
thanks @gaetanlop gonna start having a look. So perhaps you can also add something about this in the `dpo.mdx` docs too? Also if I understand now both model and ref models need to be either seq-2-seq or decoder only but not mixed?

**Comment by gaetanlop** [comment]
Hey @kashif, yes good idea for the dpo.mdx docs. I am still unsure of how we should detect if a model_init is an encoder-decoder. I don't think there is a straightforward way to do it.
Exact, if I remember well in the DPO paper the model and the ref_model are the same model. Making it mixed will require further changes. Are you aware of a recent paper using DPO with different model and ref model architectures?

**Comment by younesbelkada** [comment]
Hi @gaetanlop 
Thanks for the great work on this PR! Do you need any help finishing it up? 🙏 Let us know if you need any assistance!

**Comment by gaetanlop** [comment]
Hi @younesbelkada, sorry have been busy with other things lately. I need to add information to the dpo.mdx and resolve different conflicts. I will do this over the weekend!

**Comment by lvwerra** [comment]
Merge conflicts will be resolved probably if you merge `main` into your branch.

**Comment by gaetanlop** [comment]
Hello @younesbelkada @lvwerra, if I do not miss anything I think it is ready for a review

**Review by younesbelkada — APPROVED** [review]
This looks great to me, thanks for your great work on adding Seq2Seq model support for DPO!
